services/upload/src/main/java/app/coronawarn/server/services/federation/upload/payload/DiagnosisKeyBatchAssembler.java

```diff
@@ -31,14 +44,53 @@
    */
   public List<DiagnosisKeyBatch> assembleDiagnosisKeyBatch(List<DiagnosisKey> diagnosisKeys) {
     if (diagnosisKeys.isEmpty()) {
+      logger.info("Batches not generated: no pending upload diagnosis keys found.");
       return Collections.emptyList();
     }
-    return List.of(DiagnosisKeyBatch.newBuilder()
-        .addAllKeys(diagnosisKeys.stream()
-            .filter(DiagnosisKey::isConsentToFederation)
-            .map(this::convertKey)
-            .collect(Collectors.toList()))
-        .build());
+    if (diagnosisKeys.size() < uploadConfig.getMinBatchKeyCount()) {
+      logger.info("Batches not generated: less then minimum {} pending upload diagnosis keys.",
+          uploadConfig.getMinBatchKeyCount());
+      return Collections.emptyList();
+    }
+
+    return partionIntoBatches(filterAndConvertToUploadStructure(diagnosisKeys));
+  }
+
+  private List<DiagnosisKeyBatch>  partionIntoBatches(
+      List<app.coronawarn.server.common.protocols.external.exposurenotification.DiagnosisKey> keysToUpload) {
+
+    return  partitionListBySize(keysToUpload, uploadConfig.getMaxBatchKeyCount()).stream()
+                              .map(partition -> DiagnosisKeyBatch.newBuilder().addAllKeys(partition).build())
```

Comment:
Readability suggestion, you could extract the creation of `DiagnosisKeyBatch` to another method and call something like:
```java
.map(this::makeKeyFromPartition)
```